crates/vim/src/replace.rs

```diff
@@ -0,0 +1,174 @@
+use std::sync::Arc;
+
+use crate::{
+    motion::Motion,
+    state::{Mode, Operator},
+    Vim,
+};
+use editor::{movement, Bias};
+use gpui::{actions, ViewContext, WindowContext};
+use log::error;
+use workspace::Workspace;
+
+actions!(vim, [ToggleReplace]);
+
+pub fn register(workspace: &mut Workspace, _: &mut ViewContext<Workspace>) {
+    workspace.register_action(|_, _: &ToggleReplace, cx: &mut ViewContext<Workspace>| {
+        Vim::update(cx, |vim, cx| {
+            if vim.state().mode == Mode::Replace {
+                vim.switch_mode(Mode::Normal, false, cx);
+            } else {
+                vim.switch_mode(Mode::Replace, false, cx);
+                vim.update_active_editor(cx, |_, editor, cx| {
+                    editor.set_last_snapshot(Some(editor.buffer().clone().read(cx).snapshot(cx)));
+                });
+            }
+        });
+    });
+}
+
+pub fn replace_motion(
+    motion: Motion,
+    operator: Option<Operator>,
+    times: Option<usize>,
+    cx: &mut WindowContext,
+) {
+    Vim::update(cx, |vim, cx| {
+        match operator {
+            None => undo_replace_motion(vim, motion, times, cx),
+            Some(operator) => {
+                // Can't do anything for text objects, Ignoring
+                error!("Unexpected replace mode motion operator: {:?}", operator)
+            }
+        }
+    });
+}
+
+pub(crate) fn multi_replace(text: Arc<str>, cx: &mut WindowContext) {
+    Vim::update(cx, |vim, cx| {
+        vim.stop_recording();
+        vim.update_active_editor(cx, |_, editor, cx| {
+            editor.transact(cx, |editor, cx| {
+                editor.set_clip_at_line_ends(false, cx);
+                let (map, display_selections) = editor.selections.all_display(cx);
+                let stable_anchors = editor
+                    .selections
+                    .disjoint_anchors()
+                    .into_iter()
+                    .map(|selection| {
+                        let start = selection.start.bias_right(&map.buffer_snapshot);
+                        start..start
+                    })
+                    .collect::<Vec<_>>();
+
+                let edits = display_selections
+                    .into_iter()
+                    .map(|selection| {
+                        let mut range = selection.range();
+                        // "\n" need to be handled separately, because when a "\n" is typing,
+                        // we don't do a replace, we need insert a "\n"
+                        if text.as_ref() != "\n" {
```

Comment:
Does mapping `"enter": "editor::Newline"` in the keymap do the right thing? If not quite we may still want to call `editor.newline` as that also handles preserving leading indentation/comments etc.